### PR TITLE
docs: Document page tables, visibility, per-CPU guards

### PR DESCRIPTION
**Pull request by Zildj1an**
Document memory management of page tables, page visibility and per-CPU mapping guards. This is related to issue #74 Document COCONUT-SVSM and continues the mm documentation efforts that started with the merged [PR 243](https://github.com/coconut-svsm/svsm/pull/243). For the purpose of that issue, I will consider the memory subsystem sufficiently documented after this PR (even if not every single detail has been documented) and would like to shift focus to documenting `protocols`.

### CODE REVIEW

File: kernel/src/mm/pagetable.rs [hunk 1]
```diff
@@ -214,23 +242,28 @@ pub enum Mapping<'a> {
...
 
+    /// Get the CR3 register value for the current page table.
     pub fn cr3_value(&self) -> PhysAddr {
         let pgtable = VirtAddr::from(self as *const PageTable);
         let cr3 = virt_to_phys(pgtable);
         set_c_bit(cr3)
     }
 
+    /// Clone the shared part of the page table; excluding the private
+    /// parts.
```
Comment (stefano-garzarella) [inline kernel/src/mm/pagetable.rs:266]:
```suggestion
    /// parts.
    ///
    /// # Errors
    /// Returns `SvsmError` if the page cannot be allocated.
    ///
```

File: kernel/src/mm/pagetable.rs [hunk 2]
```diff
@@ -296,6 +365,14 @@ impl PageTable {
         };
     }
 
+    /// Walks the page table to find a mapping for a given virtual address.
+    ///
+    /// # Parameters
```
Comment (stefano-garzarella) [inline kernel/src/mm/pagetable.rs:370]:
```suggestion
    /// # Parameters
    /// - `page`: A mutable reference to the root page table.
```

File: kernel/src/mm/pagetable.rs [hunk 3]
```diff
@@ -268,12 +326,22 @@ impl PageTable {
...
+    /// # Returns
+    /// A `Mapping` representing the found mapping.
+    ///
     fn walk_addr_lvl0(page: &mut PTPage, vaddr: VirtAddr) -> Mapping<'_> {
         let idx = PageTable::index::<0>(vaddr);
 
         Mapping::Level0(&mut page[idx])
     }
 
+    /// Walks a page table at level 1 to find a mapping.
```
Comment (stefano-garzarella) [inline kernel/src/mm/pagetable.rs:344]: Should we add `# Parameters` and `# Returns` section also here, or is it redundant?

Ditto for `walk_addr_lvl2`

File: kernel/src/mm/pagetable.rs [hunk 4]
```diff
@@ -307,6 +384,7 @@ impl PageTable {
         };
     }
 
+    /// Walk the virtual address and return the corresponding mapping.
```
Comment (stefano-garzarella) [inline kernel/src/mm/pagetable.rs:387]:
```suggestion
    /// Walk the virtual address and return the corresponding mapping.
    ///
    /// # Parameters
    /// - `vaddr`: The virtual address to find a mapping for.
    ///
    /// # Returns
    /// A `Mapping` representing the found mapping.
    ///
```

**Comment by 00xc** [comment]
The new CI changes in #275 will complain about trailing `///` with no comments, i.e.

``` rust
/// Some docs
///
fn my_function() {
     ...
}
```

File: kernel/src/mm/pagetable.rs (hunk 3 above)
Comment (Zildj1an) [reply kernel/src/mm/pagetable.rs:344]: I think it may be redundant in this instance

**Comment by Zildj1an** [comment]
> The new CI changes in #275 will complain about trailing `///` with no comments, i.e.
> 
> ```rust
> /// Some docs
> ///
> fn my_function() {
>      ...
> }
> ```

Wouldn't that trigger numerous complaints for already committed comments? Also, does this imply that a block of documentation like the following:

```rust
/// Unmaps a 2MiB page from the page table sub-tree
///
/// # Arguments
///
/// * `vaddr` - The virtual address to unmap. Must be aligned to 2MiB.
///
/// # Returns
///
/// Returns a copy of the PTEntry that mapped the virtual address, if any.
///
/// # Panics
///
/// This method panics when `vaddr` is not aligned to 2MiB.
```

we will have to change it to:

```
/// Unmaps a 2MiB page from the page table sub-tree
/// # Arguments
/// * `vaddr` - The virtual address to unmap. Must be aligned to 2MiB.
/// # Returns
/// Returns a copy of the PTEntry that mapped the virtual address, if any.
/// # Panics
/// This method panics when `vaddr` is not aligned to 2MiB.
```

if so, this makes the documentation MUCH less readable from source, IMO. And what is the benefit?

**Comment by stefano-garzarella** [comment]
@Zildj1an IIUC it's just the last line of the comment block that should not be empty

**Comment by 00xc** [comment]
> > The new CI changes in #275 will complain about trailing `///` with no comments, i.e.
> > ```rust
> > /// Some docs
> > ///
> > fn my_function() {
> >      ...
> > }
> > ```
> 
> Wouldn't that trigger numerous complaints for already committed comments?

Yes, there is already a commit fixing those cases in the PR introducing the check.

> Also, does this imply that a block of documentation like the following:
> 
> ```rust
> /// Unmaps a 2MiB page from the page table sub-tree
> ///
> /// # Arguments
> ///
> /// * `vaddr` - The virtual address to unmap. Must be aligned to 2MiB.
> ///
> /// # Returns
> ///
> /// Returns a copy of the PTEntry that mapped the virtual address, if any.
> ///
> /// # Panics
> ///
> /// This method panics when `vaddr` is not aligned to 2MiB.
> ```
> 
> we will have to change it to:
> 
> ```
> /// Unmaps a 2MiB page from the page table sub-tree
> /// # Arguments
> /// * `vaddr` - The virtual address to unmap. Must be aligned to 2MiB.
> /// # Returns
> /// Returns a copy of the PTEntry that mapped the virtual address, if any.
> /// # Panics
> /// This method panics when `vaddr` is not aligned to 2MiB.
> ```
> 
> or is this just for the last line in a documentation block? If not, this makes the documentation MUCH less readable from source, IMO. And what is the benefit?

As Stefano points out, this is only for the last line :)

File: kernel/src/mm/pagetable.rs [hunk 5]
```diff
@@ -756,6 +1068,10 @@ impl RawPageTablePart {
         }
     }
 
+    /// Frees a level 2 page table, including all level 1 tables beneath it.
+    ///
+    /// # Parameters
+    /// - `page`: A reference to the level 2 page table to be freed.
     fn free_lvl2(page: &PTPage) {
```
Comment (00xc) [inline kernel/src/mm/pagetable.rs:1075]: I'd say `# Parameters` is overexplaining here.

File: kernel/src/mm/pagetable.rs [hunk 6]
```diff
@@ -746,6 +1054,10 @@ impl RawPageTablePart {
         Some(unsafe { &mut *address.as_mut_ptr::<PTPage>() })
     }
 
+    /// Frees a level 1 page table.
+    ///
+    /// # Parameters
+    /// - `page`: A reference to the level 1 page table to be freed.
```
Comment (00xc) [inline kernel/src/mm/pagetable.rs:1060]: I'd say `# Parameters` is overexplaining here.

File: kernel/src/mm/pagetable.rs [hunk 7]
```diff
@@ -663,6 +911,11 @@ impl PageTable {
         }
     }
 
+    /// Populates a part of the page table with the specified attributes.
+    ///
+    /// # Parameters
+    /// - `part`: The part of the page table to populate.
+    ///
```
Comment (00xc) [inline kernel/src/mm/pagetable.rs:918]:
```suggestion
    /// Populates this page table with the contents of the given subtree in `part`.
```

File: kernel/src/mm/pagetable.rs [hunk 8]
```diff
@@ -640,6 +883,11 @@ impl PageTable {
         Ok(())
     }
 
+    /// Unmaps a memory region.
+    ///
+    /// # Parameters
+    /// - `vregion`: The virtual memory region to unmap.
+    ///
```
Comment (00xc) [inline kernel/src/mm/pagetable.rs:890]:
```suggestion
    /// Unmaps the virtual memory region `vregion`.
```

File: kernel/src/mm/pagetable.rs [hunk 9]
```diff
@@ -605,12 +833,27 @@ impl PageTable {
         Ok(())
     }
 
+    /// Unmaps a region of memory using 2MB pages.
+    ///
+    /// # Parameters
+    /// - `vregion`: The virtual memory region to unmap.
+    ///
```
Comment (00xc) [inline kernel/src/mm/pagetable.rs:840]:
```suggestion
    /// Unmaps a region `vregion` of 2MB pages. The region must be
    /// 2MB-aligned and correspond to a set of huge mappings.
```

File: kernel/src/mm/pagetable.rs [hunk 10]
```diff
@@ -468,6 +607,17 @@ impl PageTable {
...
+    /// # Parameters
+    /// - `vaddr`: The virtual address of the page.
+    ///
+    /// # Returns
+    /// A result indicating success or an error code.
+    ///
+    /// # Errors
+    /// Returns `SvsmError` if the operation fails.
+    ///
```
Comment (00xc) [inline kernel/src/mm/pagetable.rs:620]:
```suggestion
    /// Attempts to set the C-bit for the given page at `vaddr`.
```

File: kernel/src/mm/ptguards.rs [hunk 11]
```diff
@@ -16,6 +16,7 @@ use crate::types::{PAGE_SIZE, PAGE_SIZE_2M};
 
 use crate::utils::MemoryRegion;
 
+/// Guard for a per-CPU page mapping to ensure adequate cleanup if drop.
```
Comment (00xc) [inline kernel/src/mm/ptguards.rs:19]: "if drop" sounds a bit strange. Should be "when dropped":

```suggestion
/// Guard used to unmap a per-CPU mapping when dropped.
```

File: kernel/src/mm/ptguards.rs [hunk 12]
```diff
@@ -24,6 +25,19 @@ pub struct PerCPUPageMappingGuard {
...
+    /// # Arguments
+    ///
+    /// * `paddr_start` - The starting physical address of the range.
+    /// * `paddr_end` - The ending physical address of the range.
+    /// * `alignment` - The desired alignment for the mapping.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing the [`PerCPUPageMappingGuard`] if successful,
+    /// or an `SvsmError` if an error occurs.
```
Comment (00xc) [inline kernel/src/mm/ptguards.rs:40]: Missing `# Panics` section.

File: kernel/src/mm/ptguards.rs [hunk 13]
```diff
@@ -66,10 +80,22 @@ impl PerCPUPageMappingGuard {
...
+    /// specified physical address.
+    ///
+    /// # Arguments
+    ///
+    /// * `paddr` - The physical address for the 4KB page.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing the [`PerCPUPageMappingGuard`] if successful,
+    /// or an `SvsmError` if an error occurs.
```
Comment (00xc) [inline kernel/src/mm/ptguards.rs:93]:
```suggestion
    /// Creates a new [`PerCPUPageMappingGuard`] for a 4KB page at the
    /// specified physical address, or an `SvsmError` if an error occurs.
```

File: kernel/src/mm/ptguards.rs [hunk 14]
```diff
@@ -66,10 +80,22 @@ impl PerCPUPageMappingGuard {
...
     pub fn create_4k(paddr: PhysAddr) -> Result<Self, SvsmError> {
         Self::create(paddr, paddr + PAGE_SIZE, 0)
     }
 
+    /// Returns the virtual address associated with the guard.
```
Comment (00xc) [inline kernel/src/mm/ptguards.rs:98]: This sounds like it's the address of the guard object.

```suggestion
    /// Returns the virtual address for the new mapping.
```

File: kernel/src/mm/pagetable.rs [hunk 15]
```diff
@@ -245,19 +282,44 @@ impl PageTable {
...
+    /// # Errors
+    /// Returns `SvsmError` if the page cannot be allocated.
+    ///
     fn allocate_page_table() -> Result<*mut PTPage, SvsmError> {
         let ptr = allocate_zeroed_page()?;
         Ok(ptr.as_mut_ptr::<PTPage>())
     }
 
+    /// Computes the index within a page table at the given level for a
+    /// virtual address `vaddr`.
```
[the remark this answers is not in the document]
Comment (Zildj1an) [reply kernel/src/mm/pagetable.rs:301]: Good point!

File: kernel/src/mm/pagetable.rs [hunk 16]
```diff
@@ -431,6 +548,17 @@ impl PageTable {
...
+    ///
+    /// # Parameters
+    /// - `mapping`: The mapping to split.
+    ///
+    /// # Returns
+    /// A result indicating success or an error code.
+    ///
+    /// # Errors
+    /// Returns `SvsmError` if the operation fails.
+    ///
```
[the remark this answers is not in the document]
Comment (Zildj1an) [reply kernel/src/mm/pagetable.rs:561]: In this case, I see value in documenting that the return value may be `SvsmError`. Also, the use of `mapping` as a parameter is not entirely self-evident, for example, an alternative approach could involve maintaining an index of mappings. What do you think? Let's agree on something before I force push changes

File: kernel/src/mm/pagetable.rs [hunk 17]
```diff
@@ -456,6 +584,17 @@ impl PageTable {
...
+    /// # Parameters
+    /// - `vaddr`: The virtual address of the page.
+    ///
+    /// # Returns
+    /// A result indicating success or an error code.
+    ///
+    /// # Errors
+    /// Returns `SvsmError` if the operation fails.
+    ///
     pub fn set_shared_4k(&mut self, vaddr: VirtAddr) -> Result<(), SvsmError> {
```
[the remark this answers is not in the document]
Comment (Zildj1an) [reply kernel/src/mm/pagetable.rs:598]: Same answer

File: kernel/src/mm/pagetable.rs [hunk 18]
```diff
@@ -480,6 +630,14 @@ impl PageTable {
...
     }
 
+    /// Checks the mapping of a given virtual address.
+    ///
+    /// # Parameters
+    /// - `vaddr`: The virtual address to check.
+    ///
+    /// # Returns
+    /// An option containing the physical address if a mapping exists, or `None`.
+    ///
```
[the remark this answers is not in the document]
Comment (Zildj1an) [reply kernel/src/mm/pagetable.rs:640]: Same answer

File: kernel/src/mm/pagetable.rs (hunk 10 above)
Comment (Zildj1an) [reply kernel/src/mm/pagetable.rs:620]: Same answer

File: kernel/src/mm/pagetable.rs (hunk 9 above)
Comment (Zildj1an) [reply kernel/src/mm/pagetable.rs:840]: Same answer

File: kernel/src/mm/pagetable.rs (hunk 8 above)
Comment (Zildj1an) [reply kernel/src/mm/pagetable.rs:890]: Same answer

File: kernel/src/mm/pagetable.rs (hunk 7 above)
Comment (Zildj1an) [reply kernel/src/mm/pagetable.rs:918]: Same answer

File: kernel/src/mm/pagetable.rs [hunk 19]
```diff
@@ -677,32 +930,65 @@ impl PageTable {
...
     assert!(!init_pgtable.is_set());
     *init_pgtable = pgtable;
 }
 
+/// Acquires a lock and returns a guard for the initial page table, which
+/// is locked for the duration of the guard's scope.
+///
+/// # Returns
+/// A `LockGuard` for the initial page table.
+///
```
[the remark this answers is not in the document]
Comment (Zildj1an) [reply kernel/src/mm/pagetable.rs:952]: True! I'll change it for Acquires a lock and returns a `LockGuard`

File: kernel/src/mm/pagetable.rs (hunk 6 above)
Comment (Zildj1an) [reply kernel/src/mm/pagetable.rs:1060]: Same answer. I'm leaving other comments afterwards regarding these documentation is needed or not, which seems to be something we have to figure out

File: kernel/src/mm/pagetable.rs [hunk 20]
```diff
@@ -767,18 +1083,40 @@ impl RawPageTablePart {
...
+    /// Allocates a 4KB page table entry for a given virtual address.
+    ///
+    /// # Parameters
+    /// - `vaddr`: The virtual address for which to allocate the PTE.
+    ///
+    /// # Returns
+    /// The [`Mapping`] representing the allocated or existing PTE for the address.
+    ///
+    /// # Panics
+    /// Panics if a level 3 mapping is attempted in a [`RawPageTablePart`].
```
[the remark this answers is not in the document]
Comment (Zildj1an) [reply kernel/src/mm/pagetable.rs:1119]: Agreed

File: kernel/src/mm/pagetable.rs [hunk 21]
```diff
@@ -865,6 +1238,13 @@ impl RawPageTablePart {
...
 
+    /// Unmaps a 2MB page.
+    ///
+    /// # Parameters
+    /// - `vaddr`: The virtual address of the mapping to unmap.
+    ///
+    /// # Returns
+    /// An optional [`PTEntry`] representing the unmapped page table entry.
     pub fn unmap_2m(&mut self, vaddr: VirtAddr) -> Option<PTEntry> {
         assert!(vaddr.is_aligned(PAGE_SIZE_2M));
```
[the remark this answers is not in the document]
Comment (Zildj1an) [reply kernel/src/mm/pagetable.rs:1249]: Same question

File: kernel/src/mm/ptguards.rs (hunk 12 above)
Comment (Zildj1an) [reply kernel/src/mm/ptguards.rs:40]: I don't follow, are you referring to the `assert!()`? Maybe the misunderstanding here is that I don't think we need to mark as panic a function that calls a function that may itself panic

File: kernel/src/mm/ptguards.rs (hunk 13 above)
Comment (Zildj1an) [reply kernel/src/mm/ptguards.rs:93]: Same answer

**Comment by Zildj1an** [comment]
> Left some suggestions. Overall I'd say we don't need to be too verbose adding `Arguments`, `Returns` and `Errors` sections on simple or self-explanatory functions. Also, I still see several leftover blank lines at the end of documentation comments.

I've already left a comment about this; IMHO documenting return values, potential errors and arguments is positive. The function name might be self-evident, e.g. "use A to do B" but whether the parameter to pass is "A" or,  e.g. , its index on the alphabet is not self-evident. At most it is obvious for the current contributors but not necessarily for new ones; which are the target audience of this documentation. Also, markers like '# Returns' or '# Arguments' can help non-human eyes, CI, automated tools, etc. An example is Sphinx and the rest of tools we use for the Linux kernel documentation.

File: kernel/src/mm/pagetable.rs [hunk 22]
```diff
@@ -844,6 +1207,16 @@ impl RawPageTablePart {
...
+    ///
+    /// # Parameters
+    /// - `vaddr`: The virtual address to map.
+    /// - `paddr`: The physical address to map to.
+    /// - `flags`: The flags to apply to the mapping.
+    /// - `shared`: Indicates whether the mapping is shared
+    ///
+    /// # Returns
+    /// A result indicating success (`Ok`) or failure (`Err`).
     pub fn map_2m(
```
[the remark this answers is not in the document]
Comment (00xc) [reply kernel/src/mm/pagetable.rs:1220]: If `vaddr` or `paddr` are not 2M-aligned

File: kernel/src/mm/pagetable.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (00xc) [reply kernel/src/mm/pagetable.rs:1249]: Same thing, if `vaddr` is not aligned.

File: kernel/src/mm/ptguards.rs (hunk 12 above)
Comment (00xc) [reply kernel/src/mm/ptguards.rs:40]: I'd say this panics directly, `assert!` is simply a macro that evaluates to something like `if !predicate { panic!() }`, it's the same as calling `unwrap()`

**Comment by 00xc** [comment]
> > Left some suggestions. Overall I'd say we don't need to be too verbose adding `Arguments`, `Returns` and `Errors` sections on simple or self-explanatory functions. Also, I still see several leftover blank lines at the end of documentation comments.
> 
> I've already left a comment about this; IMHO documenting return values, potential errors and arguments is positive. The function name might be self-evident, e.g. "use A to do B" but whether the parameter to pass is "A" or, e.g. , its index on the alphabet is not self-evident.

Yes, but a lot of the cases I pointed out are something like:

```rust
/// # Returns
///
/// Returns a [`Thing`], or an [`SvsmError`] if an error occured.
fn get_thing() -> Result<Thing, SvsmError> {
```

Which does not really add anything over the function signature as, well, that's the use of `Result`. So, unless we are detailing which conditions result in an error, I'd say this is not really needed.

> At most it is obvious for the current contributors but not necessarily for new ones; which are the target audience of this documentation. Also, markers like '# Returns' or '# Arguments' can help non-human eyes, CI, automated tools, etc. An example is Sphinx and the rest of tools we use for the Linux kernel documentation.

As far as I know, the use of a lot of these tools is to generate documentation, which `cargo doc` already does. For anything more advanced, like IDEs and LSPs, there is rust-analyzer. I'm not aware of any tooling in the Rust ecosystem that makes use of these sections, other than `# Safety` and `# Panics`.

File: kernel/src/mm/pagetable.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (Zildj1an) [reply kernel/src/mm/pagetable.rs:561]: Sounds good, keeping only #Returns

**Comment by Zildj1an** [comment]
I'm opening a new PR for this, with the appropriate changes following @00xc feedback

**Pull request closed without merging**